blog/_posts/2023-10-19-parallelSeriesEpisode1.md

```diff
@@ -0,0 +1,308 @@
+
+The non-linear scaling of execution time can be primarily attributed to the nature of the prime search algorithm. This algorithm involves a for loop that iterates with the size of the input number. As the input size increases, the number of iterations required by the algorithm grows proportionally. Consequently, the execution time exhibits a non-linear increase, making the algorithm less efficient for larger input sizes.
+
+While hardware constraints play a role in overall execution time, the algorithm's inherent complexity and its dependence on input size are the dominant factors in this context. Understanding the algorithm's behaviour is the first step in identifying areas for improvement.
+
+In the subsequent sections, we will explore how parallelism can address these challenges and significantly improve the performance of prime number generation in Python.
+
+#### Parallel Plain Approach
+
+We then explored a plain parallel approach, distributing the prime number search across multiple processes. The results showed promising improvements:
+
+- For 10^6 numbers, the parallel plain approach took approximately 2.41 seconds, making it approximately 1.19 times faster than the serial version.
+- With 10^7 numbers, the execution time reduced to approximately 18.56 seconds, achieving a speedup of about 3.82 times.
+- For 10^8 numbers, we observed a substantial reduction in execution time, with approximately 4 minutes (238.96 seconds), making it approximately 8.14 times faster than the serial approach.
+
+Although we have attempted to divide the workload into 40 cores theoretically, the improvement in simulation time is not significant. This may be due to the nature of the problem as checking whether a number is prime or not is not a time-consuming task. The lower performance improvement can be attributed to the substantial overhead introduced by parallelism, particularly the creation and management of multiple processes and communication between them. Often, the speed at which processes complete their tasks can surpass the efficiency of communication between cores. This leads to underutilization of available CPU cores. The image below depicts this finding: we never managed to have 40 cores using 100% of CPU for Python tasks.
+
+![Plain Parallel CPU Usage](/images/blog/parallelSeries/episode1/cpu_plain.png)
+
+#### Parallel Chunk Approach
+
+Next, we implemented a parallel chunk approach, where we divided the range of numbers into chunks and assigned each chunk to a separate process. This is an attempt to create larger process to avoid the communication overhead. The task here is find the right ballance between the time we loose managing all process and keeping all cores using 100% of CPU as much as possible. This approach allowed us to fine-tune the level of parallelism. The results for 40 chunks were particularly impressive:
```

Comment:
Time we *lose managing all processes*

Reply:
```suggestion
Next, we implemented a parallel chunk approach, where we divided the range of numbers into chunks and assigned each chunk to a separate process. This is an attempt to create larger process to avoid the communication overhead. The task here is find the right ballance between the time we lose managing all processes and keeping all cores using 100% of CPU as much as possible. This approach allowed us to fine-tune the level of parallelism. The results for 40 chunks were particularly impressive:
```